### PR TITLE
[EXPORTABLE_CODE] Update structure of package

### PR DESCRIPTION
**Pull request by akorobeinikov**

### CODE REVIEW

**Comment by akorobeinikov** [comment]
run ote_sdk tests

File: ote_sdk/ote_sdk/usecases/exportable_code/demo/README.md [hunk 1]
```diff
@@ -7,29 +7,25 @@ Demo package contains simple demo to get and visualize result of model inference
 * model
   - `model.xml`
   - `model.bin`
+  - `config.json`
 * python
   - `README.md`
   - `demo.py`
+  - `model.py (Optional)`
```
Comment (druzhkov-paul) [inline ote_sdk/ote_sdk/usecases/exportable_code/demo/README.md]:
```suggestion
  - `model.py` (Optional)
```

File: ote_sdk/ote_sdk/usecases/exportable_code/demo/README.md [hunk 2]
```diff
@@ -7,29 +7,25 @@ Demo package contains simple demo to get and visualize result of model inference
...
   - `requirements.txt`
-  - `demo_package-0.0-py3-none-any.whl`
 
+> **NOTE**: zip archieve will contain `model.py` when [ModelAPI](https://github.com/openvinotoolkit/open_model_zoo/tree/master/demos/common/python/openvino/model_zoo/model_api) haven't appropriate model wrapper for using model
```
Comment (druzhkov-paul) [inline ote_sdk/ote_sdk/usecases/exportable_code/demo/README.md]:
```suggestion
> **NOTE**: zip archive will contain `model.py` when [ModelAPI](https://github.com/openvinotoolkit/open_model_zoo/tree/master/demos/common/python/openvino/model_zoo/model_api) has no appropriate standard model wrapper for the model
```

**Comment by akorobeinikov** [comment]
run ote_sdk tests

**Comment by nervana-ff** [comment]
Can one of the admins verify this patch?

**Comment by akorobeinikov** [comment]
BMM 27040

File: external/deep-object-reid/requirements.txt [hunk 3]
```diff
@@ -1,4 +1,4 @@
 nncf @ git+https://github.com/openvinotoolkit/nncf@464244204fc2c5e80c8164c17d8d266ccae50062#egg=nncf
 openvino==2022.1.0.dev20220302
 openvino-dev==2022.1.0.dev20220302
-openmodelzoo-modelapi @ git+https://github.com/openvinotoolkit/open_model_zoo/@releases/2021/SCMVP#egg=openmodelzoo-modelapi&subdirectory=demos/common/python
\ No newline at end of file
+openmodelzoo-modelapi @ git+https://github.com/openvinotoolkit/open_model_zoo/@master#egg=openmodelzoo-modelapi&subdirectory=demos/common/python
```
Comment (Ilya-Krylov) [inline external/deep-object-reid/requirements.txt:4]: ote-related stuff has been moved to training_extensions, see external/deep-object-reid

File: external/anomaly/requirements.txt [hunk 4]
```diff
@@ -1,5 +1,5 @@
 anomalib @ git+https://github.com/openvinotoolkit/anomalib.git@5f3ee2725d97af8a0a7865b2fcac7280140bfc08
-openmodelzoo-modelapi @ git+https://github.com/openvinotoolkit/open_model_zoo/@releases/2021/SCMVP#egg=openmodelzoo-modelapi&subdirectory=demos/common/python
+openmodelzoo-modelapi @ git+https://github.com/openvinotoolkit/open_model_zoo/@master#egg=openmodelzoo-modelapi&subdirectory=demos/common/python
```
Comment (druzhkov-paul) [inline external/anomaly/requirements.txt]: I guess it should be later fixed to some fixed/frozen release state. For now it is Ok, but don't forget about it moving forward.

File: ote_sdk/ote_sdk/entities/shapes/ellipse.py [hunk 5]
```diff
@@ -277,3 +277,9 @@ def get_area(self) -> float:
         :return: area of the shape
         """
         return math.pi * self.minor_axis * self.major_axis
+
+    def to_rectangle(self) -> Rectangle:
```
Comment (druzhkov-paul) [inline ote_sdk/ote_sdk/entities/shapes/ellipse.py]: `ShapeFactory.shape_as_rectangle` ?

File: ote_sdk/ote_sdk/entities/shapes/rectangle.py [hunk 6]
```diff
@@ -90,7 +90,6 @@ def __eq__(self, other):
                 and self.y1 == other.y1
                 and self.x2 == other.x2
                 and self.y2 == other.y2
-                and self.modification_date == other.modification_date
```
Comment (druzhkov-paul) [inline ote_sdk/ote_sdk/entities/shapes/rectangle.py]: Please elaborate.

File: ote_sdk/ote_sdk/usecases/exportable_code/demo/README.md [hunk 7]
```diff
@@ -53,74 +57,88 @@ Demo package contains simple demo to get and visualize result of model inference
...
    ```
 
-
-When the package is installed, you can import it as follows:
-```
-python -c "from demo_package import create_model"
-```
+4. Add `model_wrappers` package to PYTHONPATH:
+   ```
+   export PYTHONPATH=$PYTHONPATH:/path/to/model_wrappers
```
Comment (druzhkov-paul) [inline ote_sdk/ote_sdk/usecases/exportable_code/demo/README.md:69]: What about Windows shells?

File: ote_sdk/ote_sdk/usecases/exportable_code/demo/README.md [hunk 8]
```diff
@@ -53,74 +57,88 @@ Demo package contains simple demo to get and visualize result of model inference
...
 
    ```
 
-   As a model, you can use `model.xml` from generated zip. So can use the following command to do inference with a pre-trained model:
+   As a model, you can use path to model directory from generated zip. So you can use the following command to do inference with a pre-trained model:
    ```
    python3 demo.py \
      -i <path_to_video>/inputVideo.mp4 \
-     -m <path_to_model>/model.xml
+     -m <path_to_model_directory> \
```
Comment (druzhkov-paul) [inline ote_sdk/ote_sdk/usecases/exportable_code/demo/README.md]: ?

File: ote_sdk/ote_sdk/usecases/exportable_code/demo/README.md [hunk 9]
```diff
@@ -53,74 +57,88 @@ Demo package contains simple demo to get and visualize result of model inference
...
+   > **NOTE**: If you provide a single image as an input, the demo processes and renders it quickly, then exits. To continuously
+   > visualize inference results on the screen, apply the `loop` option, which enforces processing a single image in a loop.
+
    > **NOTE**: Default configuration contains info about pre- and postprocessing to model inference and is guaranteed to be correct.
-   > Also you can define own json config that specifies needed parameters, but any change should be made with caution.
-   > To create this config please see `config.json` in demo_package wheel.
+   > Also you can change `config.json` that specifies needed parameters, but any change should be made with caution.
 
-2. You can create your own demo application, using `demo_package`. The main function of package is `create_model`:
+2. You can create your own demo application, using `demo_package`. The main class of package is `ModelEntity`.
```
Comment (druzhkov-paul) [inline ote_sdk/ote_sdk/usecases/exportable_code/demo/README.md]: What is `demo_package`?

File: ote_sdk/ote_sdk/usecases/exportable_code/demo/README.md [hunk 10]
```diff
@@ -53,74 +57,88 @@ Demo package contains simple demo to get and visualize result of model inference
...
    ```python
-   def create_model(model_path: Path, config_file: Path = None) -> Model:
-    """
-    Create model using ModelAPI factory
-
-    :param model_path: Path to .xml model
-    :param config_file: Path to .json config. If it is not defined, use config from demo_package
-    """
+   class ModelContainer:
```
Comment (druzhkov-paul) [inline ote_sdk/ote_sdk/usecases/exportable_code/demo/README.md]: What's the difference between `ModelEntity` and `ModelContainer`?

File: ote_sdk/ote_sdk/usecases/exportable_code/demo/README.md (hunk 9 above)
Comment (druzhkov-paul) [reply ote_sdk/ote_sdk/usecases/exportable_code/demo/README.md]: Is this section about ad-hoc demo or ad-hoc model (wrapper)? If the latter, why should the user care about it at all?

File: ote_sdk/ote_sdk/usecases/exportable_code/demo/demo.py [hunk 11]
```diff
@@ -38,35 +42,75 @@ def build_argparser():
...
+def get_inferencer_class(type_inference, models):
+    """
+    Return class for inference of models
+    """
+    if type_inference == "chain" and len(models) == 1:
+        raise RuntimeError(
+            "For single model please use 'sync' or 'async' type of inference"
+        )
+    if len(models) > 1:
+        type_inference = "chain"
```
Comment (druzhkov-paul) [inline ote_sdk/ote_sdk/usecases/exportable_code/demo/demo.py:83]: Maybe add some message to notify the user about that?

File: ote_sdk/ote_sdk/usecases/exportable_code/demo/demo.py [hunk 12]
```diff
@@ -38,35 +42,75 @@ def build_argparser():
...
     """
     args = build_argparser().parse_args()
-    # create components for demo
+    # create models and converters for outputs
+    models = []
+    converters = []
+    for model_dir in args.models:
+        model = ModelContainer(model_dir)
+        models.append(model)
+        converters.append(create_output_converter(model.task_type, model.labels))
```
Comment (druzhkov-paul) [inline ote_sdk/ote_sdk/usecases/exportable_code/demo/demo.py]: Where converters are used?

File: ote_sdk/ote_sdk/usecases/exportable_code/demo/demo_package/executors/asynchronous.py [hunk 13]
```diff
@@ -0,0 +1,70 @@
...
+class AsyncExecutor:
+    """
+    Async inferencer
+
+    Args:
+        model: model for inference
+        visualizer: for visualize inference results
+    """
+
+    def __init__(self, model, visualizer) -> None:
```
Comment (druzhkov-paul) [inline ote_sdk/ote_sdk/usecases/exportable_code/demo/demo_package/executors/asynchronous.py]: Type annotation is required for SDK code.

**Comment by druzhkov-paul** [comment]
run ote_sdk tests

**Comment by druzhkov-paul** [comment]
run ote_sdk tests

File: ote_sdk/ote_sdk/entities/shapes/rectangle.py (hunk 6 above)
Comment (akorobeinikov) [reply ote_sdk/ote_sdk/entities/shapes/rectangle.py]: I need to compare parent annotation from previous stage with item annotation from current stage in task chain. If shapes are equal I don't add new annotation to result scene, I only add labels from item to parent annotation. `Modification date` don't allow doing it compare correctly. In general I think modification date shouldn't participate in equal operator.

File: ote_sdk/ote_sdk/entities/shapes/ellipse.py (hunk 5 above)
Comment (akorobeinikov) [reply ote_sdk/ote_sdk/entities/shapes/ellipse.py]: Now used factory

File: ote_sdk/ote_sdk/usecases/exportable_code/demo/README.md [hunk 14]
```diff
@@ -1,25 +1,29 @@
...
+## Structure of generated zip:
 
 * model
   - `model.xml`
   - `model.bin`
+  - `config.json`
 * python
+  - model_wrappers (Optional)
+    - `__init__.py`
+    - model_wrappers needed for run demo
```
Comment (druzhkov-paul) [inline ote_sdk/ote_sdk/usecases/exportable_code/demo/README.md]:
```suggestion
    - model_wrappers required to run demo
```

File: ote_sdk/ote_sdk/usecases/exportable_code/demo/README.md [hunk 15]
```diff
@@ -1,25 +1,29 @@
-# Exportable code - demo package
+# Exportable code
 
-Demo package contains simple demo to get and visualize result of model inference.
+Exportable code is a .zip archieve that contains simple demo to get and visualize result of model inference.
```
Comment (druzhkov-paul) [inline ote_sdk/ote_sdk/usecases/exportable_code/demo/README.md]:
```suggestion
Exportable code is a .zip archive that contains simple demo to get and visualize result of model inference.
```

File: ote_sdk/ote_sdk/usecases/exportable_code/demo/README.md [hunk 16]
```diff
@@ -1,25 +1,29 @@
...
+  - model_wrappers (Optional)
+    - `__init__.py`
+    - model_wrappers needed for run demo
   - `README.md`
   - `LICENSE`
   - `demo.py`
   - `requirements.txt`
-  - `demo_package-0.0-py3-none-any.whl`
 
+> **NOTE**: zip archive will contain model_wrappers when [ModelAPI](https://github.com/openvinotoolkit/open_model_zoo/tree/master/demos/common/python/openvino/model_zoo/model_api) has no appropriate standard model wrapper for the model
```
Comment (druzhkov-paul) [inline ote_sdk/ote_sdk/usecases/exportable_code/demo/README.md]:
```suggestion
> **NOTE**: zip archive contains model_wrappers when [ModelAPI](https://github.com/openvinotoolkit/open_model_zoo/tree/master/demos/common/python/openvino/model_zoo/model_api) has no appropriate standard model wrapper for the model
```

File: ote_sdk/ote_sdk/usecases/exportable_code/demo/README.md [hunk 17]
```diff
@@ -53,81 +57,100 @@ Demo package contains simple demo to get and visualize result of model inference
...
-python -c "from demo_package import create_model"
-```
+   On Linux and macOS:
+   ```
+   export PYTHONPATH=$PYTHONPATH:/path/to/model_wrappers
+   ```
+
+   On Windows:
+   ```
+   set PYTHONPATH=$PYTHONPATH:/path/to/model_wrappers
```
Comment (druzhkov-paul) [inline ote_sdk/ote_sdk/usecases/exportable_code/demo/README.md]: Really?

File: ote_sdk/ote_sdk/usecases/exportable_code/demo/README.md [hunk 18]
```diff
@@ -53,81 +57,100 @@ Demo package contains simple demo to get and visualize result of model inference
...
-   > To create this config please see `config.json` in demo_package wheel.
+   > **NOTE**: If you provide a single image as an input, the demo processes and renders it quickly, then exits. To continuously
+   > visualize inference results on the screen, apply the `loop` option, which enforces processing a single image in a loop.
 
-2. You can create your own demo application, using `demo_package`. The main function of package is `create_model`:
-   ```python
-   def create_model(model_path: Path, config_file: Path = None) -> Model:
-    """
-    Create model using ModelAPI factory
+   > **NOTE**: Default configuration contains info about pre- and postprocessing to model inference and is guaranteed to be correct.
```
Comment (druzhkov-paul) [inline ote_sdk/ote_sdk/usecases/exportable_code/demo/README.md]: What do you mean saying "postprocessing to model inference"?

File: ote_sdk/ote_sdk/usecases/exportable_code/demo/README.md [hunk 19]
```diff
@@ -53,81 +57,100 @@ Demo package contains simple demo to get and visualize result of model inference
...
-   # create model
-   model = create_model(path_to_model)
-   # inference
-   objects = model(frame)
-   # show results using some visualizer
-   output = visualizer.draw(frame, objects)
-   cv2.imshow(output)
    ```
 
+   > **NOTE**: Model wrappers contains pre- and postprocessing operations needed to inference. Default name of model wrapper provided in `config.json` as `type_of_model`. The wrappers themselves stored at model wrapper folder or at ModelAPI OMZ. To get more information please see [ModelAPI](https://github.com/openvinotoolkit/open_model_zoo/tree/master/demos/common/python/openvino/model_zoo/model_api). If you want to use your own model wrapper you should create wrapper in `model_wrappers` directory (if there is no this directory create it) and change `type_of_model` field in `config.json` according to wrapper.
```
Comment (druzhkov-paul) [inline ote_sdk/ote_sdk/usecases/exportable_code/demo/README.md]:
```suggestion
   > **NOTE**: Model wrappers contain pre- and postprocessing operations needed for inference. Default name of model wrapper provided in `config.json` as `type_of_model`. The wrappers themselves stored in model wrapper folder or at ModelAPI OMZ. To get more information please see [ModelAPI](https://github.com/openvinotoolkit/open_model_zoo/tree/master/demos/common/python/openvino/model_zoo/model_api). If you want to use your own model wrapper you should create wrapper in `model_wrappers` directory and change `type_of_model` field in `config.json` according to wrapper.
```

Comment (druzhkov-paul) [reply ote_sdk/ote_sdk/usecases/exportable_code/demo/README.md]: In general, it seems to me that you're trying to address two separate points in a single readme file:

1. Provide description of the generated exportable code (which this readme is a part of) archive and manual for setting it up and running.
2. Developers guide for exportable code related stuff.

If so, I'd propose to explicitly split it into two pieces and distribute only the first one along with the exportable code archive.

File: ote_sdk/ote_sdk/usecases/exportable_code/demo/README.md [hunk 20]
```diff
@@ -53,81 +57,100 @@ Demo package contains simple demo to get and visualize result of model inference
...
 1. If you have access to the Internet through the proxy server only, please use pip with proxy call as demonstrated by command below:
    ```
    python -m pip install --proxy http://<usr_name>:<password>@<proxyserver_name>:<port#> <pkg_name>
    ```
 
-2. If you use Anaconda environment, you should consider that OpenVINO has limited [Conda support](https://docs.openvino.ai/2021.4/openvino_docs_install_guides_installing_openvino_conda.html) for Python 3.6 and 3.7 versions only. But the demo package requires python 3.8. So please use other tools to create the environment (like `venv` or `virtualenv`) and use `pip` as a package manager.
\ No newline at end of file
+2. If you use Anaconda environment, you should consider that OpenVINO has limited [Conda support](https://docs.openvino.ai/2021.4/openvino_docs_install_guides_installing_openvino_conda.html) for Python 3.6 and 3.7 versions only. But the demo package requires python 3.8. So please use other tools to create the environment (like `venv` or `virtualenv`) and use `pip` as a package manager.
+
+3. If you have problems when you try yo use `pip install` command, please update pip version by following command:
```
Comment (druzhkov-paul) [inline ote_sdk/ote_sdk/usecases/exportable_code/demo/README.md]: Yo! Please correct spelling thoroughly everywhere.

File: ote_sdk/ote_sdk/usecases/exportable_code/demo/README.md [hunk 21]
```diff
@@ -53,81 +57,100 @@ Demo package contains simple demo to get and visualize result of model inference
...
-     -c CONFIG, --config CONFIG
-                           Optional. Path to an .json file with parameters for
-                           model.
+     -m MODELS [MODELS ...], --models MODELS [MODELS ...]
+                           Required. Path to directory with trained model and
+                           configuration file. For task chain please provide
+                           models-participants in right order
+     -it {sync,async,chain}, --inference_type {sync,async,chain}
+                           Optional. Type of inference. For task-chain you should
+                           type 'chain'.
```
Comment (druzhkov-paul) [inline ote_sdk/ote_sdk/usecases/exportable_code/demo/README.md]: Or should not as it is auto corrected?

File: ote_sdk/ote_sdk/usecases/exportable_code/demo/demo.py [hunk 22]
```diff
@@ -38,35 +41,75 @@ def build_argparser():
...
+    """
+    Return class for inference of models
+    """
+    if type_inference == "chain" and len(models) == 1:
+        raise RuntimeError(
+            "For single model please use 'sync' or 'async' type of inference"
+        )
+    if len(models) > 1:
+        type_inference = "chain"
+        print("You run task chain pipeline with provided models")
```
Comment (druzhkov-paul) [inline ote_sdk/ote_sdk/usecases/exportable_code/demo/demo.py]: Could be better... Why does task chain pipeline runs? What corrections of the user input are done implicitly?

File: ote_sdk/ote_sdk/usecases/exportable_code/demo/demo_package/executors/asynchronous.py [hunk 23]
```diff
@@ -0,0 +1,75 @@
...
+        visualizer: for visualize inference results
+    """
+
+    def __init__(self, model: ModelContainer, visualizer: Visualizer) -> None:
+        self.model = model.core_model
+        self.visualizer = visualizer
+        self.converter = create_output_converter(model.task_type, model.labels)
+        self.async_pipeline = AsyncPipeline(self.model)
+
+    def run(self, input_stream: Union[int, str], loop=False):
```
Comment (druzhkov-paul) [inline ote_sdk/ote_sdk/usecases/exportable_code/demo/demo_package/executors/asynchronous.py]:
```suggestion
    def run(self, input_stream: Union[int, str], loop: bool = False):
```

Please double-check everywhere.

File: ote_sdk/ote_sdk/usecases/exportable_code/demo/demo_package/executors/asynchronous.py [hunk 24]
```diff
@@ -0,0 +1,75 @@
...
+                visualizer.show(output)
+
+    def render_result(self, results: Tuple[Any, dict]):
+        """
+        Render for results of inference
+        """
+        predictions, frame_meta = results
+        annotation_scene = self.converter.convert_to_annotation(predictions, frame_meta)
+        current_frame = frame_meta["frame"]
+        # any user's visualizer
```
Comment (druzhkov-paul) [inline ote_sdk/ote_sdk/usecases/exportable_code/demo/demo_package/executors/asynchronous.py]: This comment looks irrelevant here.

File: ote_sdk/ote_sdk/usecases/exportable_code/demo/demo_package/executors/asynchronous.py [hunk 25]
```diff
@@ -0,0 +1,75 @@
...
+from ote_sdk.usecases.exportable_code.visualizers import HandlerVisualizer, Visualizer
+
+
+class AsyncExecutor:
+    """
+    Async inferencer
+
+    Args:
+        model: model for inference
+        visualizer: for visualize inference results
```
Comment (druzhkov-paul) [inline ote_sdk/ote_sdk/usecases/exportable_code/demo/demo_package/executors/asynchronous.py]:
```suggestion
        visualizer: visualizer of inference results
```

File: ote_sdk/ote_sdk/usecases/exportable_code/demo/demo_package/executors/asynchronous.py [hunk 26]
```diff
@@ -0,0 +1,75 @@
...
+                self.async_pipeline.submit_data(frame, next_frame_id, {"frame": frame})
+                next_frame_id += 1
+
+            self.async_pipeline.await_all()
+            for next_frame_id_to_show in range(next_frame_id_to_show, next_frame_id):
+                results = self.async_pipeline.get_result(next_frame_id_to_show)
+                output = self.render_result(results)
+                visualizer.show(output)
+
+    def render_result(self, results: Tuple[Any, dict]):
```
Comment (druzhkov-paul) [inline ote_sdk/ote_sdk/usecases/exportable_code/demo/demo_package/executors/asynchronous.py]: Type hint for output.

File: ote_sdk/ote_sdk/usecases/exportable_code/demo/demo_package/executors/sync_pipeline.py [hunk 27]
```diff
@@ -0,0 +1,108 @@
...
+                            parent_annotation.append_label(label)
+                    else:
+                        result_scene.append_annotation(item_annotation)
+            current_objects = new_objects
+        return result_scene
+
+    @staticmethod
+    def crop(
+        item: np.ndarray, parent_annotation: Annotation, item_annotation: Annotation
+    ):
```
Comment (druzhkov-paul) [inline ote_sdk/ote_sdk/usecases/exportable_code/demo/demo_package/executors/sync_pipeline.py]: Type hint for output.

**Comment by akorobeinikov** [comment]
BMM 27246 ✔️

File: ote_sdk/ote_sdk/usecases/exportable_code/streamer/streamer.py [hunk 28]
```diff
@@ -8,174 +8,45 @@
...
-        >>> path = Path("./demo.mp4")
-        >>> extensions = _get_extensions(media_type=MediaType.VIDEO)
-        >>> _is_file_with_supported_extensions(path, extensions)
-        True
 
-        >>> path = Path("demo.jpg")
-        >>> extensions = _get_extensions(media_type=MediaType.IMAGE)
-        >>> _is_file_with_supported_extensions(path, extensions)
-        True
+    def __init__(self, message):
```
Comment (druzhkov-paul) [inline ote_sdk/ote_sdk/usecases/exportable_code/streamer/streamer.py]: Type hints.

**Review by druzhkov-paul — CHANGES_REQUESTED** [review]

File: ote_sdk/ote_sdk/entities/shapes/rectangle.py (hunk 6 above)
Comment (druzhkov-paul) [reply ote_sdk/ote_sdk/entities/shapes/rectangle.py]: Whether modification date should be a part of the SDK at all is a separate discussion. But for now, I'd suggest you to directly work with a single annotation scene and modify it inplace. Using DatasetItem API should also help to handle different ROIs and corresponding crops.

File: ote_sdk/ote_sdk/usecases/exportable_code/demo/README.md (hunk 19 above)
Comment (akorobeinikov) [reply ote_sdk/ote_sdk/usecases/exportable_code/demo/README.md]: I deleted developer guide from this readme. I agree that it should be separated part. Maybe it make sense to add `how to` notebook, but not in this PR.

File: ote_sdk/ote_sdk/entities/shapes/rectangle.py (hunk 6 above)
Comment (akorobeinikov) [reply ote_sdk/ote_sdk/entities/shapes/rectangle.py]: `modification date` is returned. Now pipeline doesn't base on comparing of shapes

File: ote_sdk/ote_sdk/usecases/exportable_code/demo/README.md [hunk 29]
```diff
@@ -71,75 +71,46 @@ Exportable code is a .zip archieve that contains simple demo to get and visualiz
...
-   > **NOTE**: If you provide a single image as an input, the demo processes and renders it quickly, then exits. To continuously
-   > visualize inference results on the screen, apply the `loop` option, which enforces processing a single image in a loop.
+As a model, you can use path to model directory from generated zip. So you can use the following command to do inference with a pre-trained model:
 
-   > **NOTE**: Default configuration contains info about pre- and postprocessing to model inference and is guaranteed to be correct.
-   > Also you can change `config.json` that specifies needed parameters, but any change should be made with caution.
+```
+python3 demo.py \
+  -i <path_to_video>/inputVideo.mp4 \
+  -m <path_to_model> \
```
Comment (druzhkov-paul) [inline ote_sdk/ote_sdk/usecases/exportable_code/demo/README.md]:
```suggestion
  -m <path_to_model>
```

**Comment by druzhkov-paul** [comment]
run ote-test

**Comment by druzhkov-paul** [comment]
run ote_sdk tests

File: ote_sdk/ote_sdk/usecases/exportable_code/demo/demo_package/model_container.py [hunk 30]
```diff
@@ -0,0 +1,84 @@
...
+        Task type property
+        """
+        return self._task_type
+
+    @property
+    def is_global(self) -> bool:
+        """
+        Return True if the task produces global labels, False otherwise
+        """
+        return self._task_type in GLOBAL_TASK_TYPES
```
Comment (druzhkov-paul) [inline ote_sdk/ote_sdk/usecases/exportable_code/demo/demo_package/model_container.py]: `TaskType` enum from (ote_sdk/ote_sdk/entities/model_template.py) has been recently enhanced to provide some additional information about the Task, including global/local identifiers. It'd be good to reuse this bit of logic rather than introduce one more to do so. And in general `task_type` here could be `TaskType` not a string.

File: ote_sdk/ote_sdk/usecases/exportable_code/demo/demo_package/executors/sync_pipeline.py [hunk 31]
```diff
@@ -0,0 +1,106 @@
...
+    ) -> Tuple[np.ndarray, Annotation]:
+        """
+        Crop operation between chain stages
+        """
+        new_item = ShapeFactory.shape_as_rectangle(
+            item_annotation.shape
+        ).crop_numpy_array(item)
+        item_annotation.shape = item_annotation.shape.normalize_wrt_roi_shape(
+            ShapeFactory.shape_as_rectangle(parent_annotation.shape)
+        )
```
Comment (druzhkov-paul) [inline ote_sdk/ote_sdk/usecases/exportable_code/demo/demo_package/executors/sync_pipeline.py:90]: I still have a feeling that the code can be made cleaner if using `DatasetItem`s. It handles ROIs, cropping and normalization transparently. This being said, I'm Ok to leave it as is.

**Comment by druzhkov-paul** [comment]
run ote_sdk tests

**Comment by druzhkov-paul** [comment]
run ote-test

**Comment by druzhkov-paul** [comment]
run ote_sdk tests

File: external/mmsegmentation/segmentation_tasks/apis/segmentation/openvino_task.py [hunk 32]
```diff
@@ -209,47 +206,38 @@ def deploy(self,
...
-                zip.write(os.path.join(tempdir, "requirements.txt"), os.path.join("python", "requirements.txt"))
-                zip.write(os.path.join(work_dir, "README.md"), os.path.join("python", "README.md"))
-                zip.write(os.path.join(work_dir, "LICENSE"), os.path.join("python", "LICENSE"))
-                zip.write(os.path.join(work_dir, "demo.py"), os.path.join("python", "demo.py"))
-                zip.write(os.path.join(tempdir, wheel_file_name), os.path.join("python", wheel_file_name))
-            with open(os.path.join(tempdir, "openvino.zip"), "rb") as file:
-                output_model.exportable_code = file.read()
+
+        zip_buffer = io.BytesIO()
+        with ZipFile(zip_buffer, 'w') as arch:
```
Comment (sovrasov) [inline external/mmsegmentation/segmentation_tasks/apis/segmentation/openvino_task.py:216]: Seems like zip buffer is not saved later:
`output_model.exportable_code = zip_buffer.getvalue()` is missing

**Comment by druzhkov-paul** [comment]
run ote-test

**Comment by druzhkov-paul** [comment]
run ote-test

**Comment by druzhkov-paul** [comment]
Tests [failed](http://validationreports.sclab.intel.com:8006/reports/build_number_report?test_session_build_number=ote-test-1590&environment=iotg-dev-workstation-12).

**Comment by druzhkov-paul** [comment]
And test_ote_training tests [here](http://validationreports.sclab.intel.com:8006/reports/build_number_report?test_session_build_number=ote-test-1592&environment=iotg-dev-workstation-17) failed as well.

**Comment by druzhkov-paul** [comment]
run ote-test

**Comment by druzhkov-paul** [comment]
run ote_sdk tests

**Comment by aiarkinx** [comment]
run ote-test

**Comment by Ilya-Krylov** [comment]
run ote_sdk tests

**Comment by druzhkov-paul** [comment]
external/deep-object-reid/ote_cli/test_classification.py::TestToolsClassification::test_ote_eval_openvino[Custom_Image_Classification_EfficinetNet-B0] [PASSED](http://validationreports.sclab.intel.com:8006/reports/test_result/6244271853b4a08578dac6b3)

**Review by druzhkov-paul — APPROVED** [review]
